### PR TITLE
add version in the entity cache hash

### PR DESCRIPTION
**Pull request by bnjjj**
---

**Checklist**

Complete the checklist (and note appropriate exceptions) before the PR is marked ready-for-review.

- [x] Changes are compatible[^1]
- [ ] Documentation[^2] completed
- [x] Performance impact assessed and acceptable
- Tests added and passing[^3]
    - [x] Unit Tests
    - [x] Integration Tests
    - [ ] Manual Tests

**Exceptions**

*Note any exceptions here*

**Notes**

[^1]: It may be appropriate to bring upcoming changes to the attention of other (impacted) groups. Please endeavour to do this before seeking PR approval. The mechanism for doing this will vary considerably, so use your judgement as to how and when to do this.
[^2]: Configuration is an important part of many changes. Where applicable please try to document configuration examples.
[^3]: Tick whichever testing boxes are applicable. If you are adding Manual Tests, please document the manual testing (extensively) in the Exceptions.

[ROUTER-424]: https://apollographql.atlassian.net/browse/ROUTER-424?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: apollo-router/src/plugins/cache/entity.rs [hunk 1]
```diff
@@ -917,7 +919,7 @@ fn extract_cache_key_root(
     let mut key = String::new();
     let _ = write!(
         &mut key,
-        "subgraph:{subgraph_name}:type:{entity_type}:hash:{query_hash}:data:{additional_data_hash}"
+        "version:{ENTITY_CACHE_VERSION}:subgraph:{subgraph_name}:type:{entity_type}:hash:{query_hash}:data:{additional_data_hash}"
```
Comment (garypen) [inline apollo-router/src/plugins/cache/entity.rs:922]: The comments above need to be updated.

File: apollo-router/src/plugins/cache/entity.rs [hunk 2]
```diff
@@ -972,7 +974,7 @@ fn extract_cache_keys(
         // - query hash: invalidate the entry for a specific query and operation name
         // - additional data: separate cache entries depending on info like authorization status
         let mut key = String::new();
-        let _ = write!(&mut key,  "subgraph:{subgraph_name}:{typename}:{hashed_entity_key}:{query_hash}:{additional_data_hash}");
+        let _ = write!(&mut key,  "version:{ENTITY_CACHE_VERSION}:subgraph:{subgraph_name}:{typename}:{hashed_entity_key}:{query_hash}:{additional_data_hash}");
```
Comment (garypen) [inline apollo-router/src/plugins/cache/entity.rs:977]: The comments above need to be updated.

File: apollo-router/src/plugins/cache/entity.rs [hunk 3]
```diff
@@ -50,6 +50,8 @@ use crate::services::supergraph;
 use crate::spec::TYPENAME;
 use crate::Context;
 
+/// Change this key if you introduce a breaking change in entity caching algorithm to make sure it won't take the previous entries
+pub(crate) const ENTITY_CACHE_VERSION: &str = "0.1";
```
Comment (garypen) [inline apollo-router/src/plugins/cache/entity.rs:54]: It's not a hard feeling, but I feel like we want to version from `1.0`. It sounds out a more confident message.

**Review by BrynCooke — APPROVED** [review]

**Review by BrynCooke — CHANGES_REQUESTED** [review]
Needs to be called out strongly in the changelog

**Pull request merged**